lib/ruby_lsp/requests/code_action_resolve.rb

```diff
@@ -45,11 +46,44 @@ def run
         source_range = @code_action.dig(:data, :range)
         return Error::EmptySelection if source_range[:start] == source_range[:end]
 
+        return Error::InvalidTargetRange if @document.syntax_error?
+
         scanner = @document.create_scanner
         start_index = scanner.find_char_position(source_range[:start])
         end_index = scanner.find_char_position(source_range[:end])
-        extraction_source = T.must(@document.source[start_index...end_index])
-        source_line_indentation = T.must(T.must(@document.source.lines[source_range.dig(:start, :line)])[/\A */]).size
+        extracted_source = T.must(@document.source[start_index...end_index])
+
+        # Find the closest statements node, so that we place the refactor in a valid position
+        closest_statements = locate(T.must(@document.tree), start_index, node_types: [SyntaxTree::Statements]).first
+        return Error::InvalidTargetRange if closest_statements.nil?
+
+        # Find the node with the end line closest to the requested position, so that we can place the refactor
+        # immediately after that closest node
+        closest_node = closest_statements.child_nodes.compact.min_by do |node|
+          distance = source_range.dig(:start, :line) - (node.location.end_line - 1)
+          distance <= 0 ? Float::INFINITY : distance
+        end
+
+        # When trying to extract the first node inside of a statements block, then we can just select one line above it
+        target_line = if closest_node == closest_statements.child_nodes.first
+          closest_node.location.start_line - 1
+        else
+          closest_node.location.end_line
+        end
+
+        lines = @document.source.lines
+        indentation = T.must(T.must(lines[target_line - 1])[/\A */]).size
```

Comment:
To be honest, I prefer the `T.must`. The method `fetch` is used when you want to grab a value and set a default, which in my opinion is a different intention than saying "I want to grab this value and it must not be nil".

Reply:
I don't really think of `fetch` that way, for hashes I almost always use it instead of `[]` even when no default is needed. But admittedly it's not common to see it used with arrays.

![fetch](https://user-images.githubusercontent.com/13400/222567783-461c24da-b5d5-4fd5-9855-8767279b9ae7.gif)